modules/core/src/main/java/org/apache/ignite/internal/util/VersatileReadWriteLock.java

```diff
@@ -338,15 +338,15 @@ public boolean tryWriteLock(long timeout, TimeUnit unit) throws InterruptedExcep
      * @throws IllegalMonitorStateException thrown if the write lock is not acquired.
      */
     public void writeUnlock() {
-        int curState = state;
-        // There could still be some read locks if the write lock was taken forcefully.
-        int readLocks = readLocks(curState);
+        while (true) {
+            int curState = state;
+            // There could still be some read locks if the write lock was taken forcefully.
+            int readLocks = readLocks(curState);
 
-        if (!writeLocked(curState)) {
-            throw new IllegalMonitorStateException();
-        }
+            if (!writeLocked(curState)) {
+                throw new IllegalMonitorStateException();
+            }
 
-        while (true) {
             if (STATE_VH.compareAndSet(this, state, state(false, readLocks))) {
```

Comment:
Why do we use `state` here and not local `curState` instead?

Reply:
Ouch. Good catch, fixed